File: AddTwoNumbers/AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        
        iter1 = l1
        iter2 = l2
        
        head = ListNode(0)
        result = head
        
        carry = 0
        
        while iter1 or iter2 or carry:
            sum = carry
            if iter1:
                sum += iter1.val
                
            if iter2:
                sum += iter2.val
                
            carry, rem = divmod(sum, 10)
            head.next = ListNode(rem)
            
            head = head.next
            
            iter1 = iter1.next if iter1.next else None
            iter2 = iter2.next if iter2.next else None
        
        return result.next;

File: 0002-AddTwoNumbers/AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* iter1 = l1;
        ListNode* iter2 = l2;
        
        ListNode* head = new ListNode(0);
        ListNode* result = head;
        
        int carry = 0;
        
        while(l1 != nullptr || l2 != nullptr || carry){
            int sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
            head->next = new ListNode(sum % 10);
            carry = sum / 10;
            head = head->next;
            
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
        }
        
        return result->next;
    }
};
